feat: add @link operator

melsonic opened the pull request:
<!-- **Summary:**  
_Briefly describe the changes made in this PR._

**Issue Reference(s):**  
Fixes #... _(Replace "..." with the issue number)_

**Build & Testing:**

- [ ] I ran `cargo test` successfully.
- [ ] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [ ] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation] accordingly.
- [ ] I have performed a self-review of my code.
- [ ] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcall/tree/main/docs
-->

Initial draft PR
Fixes #760

tusharmath commented on src/config/resolver.rs:
```diff
@@ -0,0 +1,48 @@
+
+impl Link {
+
+  pub async fn read_link_type(&mut self) -> Result<(), anyhow::Error> {
```
```suggestion
  pub async fn resolve_recurse(self) -> Result<Link, anyhow::Error> {
```

melsonic replied:
Resolved :+1:

tusharmath commented on src/config/resolver.rs:
```diff
@@ -0,0 +1,48 @@
+      self.content = Some(String::from_utf8(buffer)?);
+    };
+    
+    Ok(())
```
Need to check for recursive links so that it doesn't get into a loop. 

- [ ] Add unit tests for this component
- [ ] Keep File and Http as a dependency so that it's easier to test.

melsonic replied:
@tusharmath Regarding the recursive links, the `resolver` module only fetches the file content and places it into the `content` attribute of the `Link`, as it was mentioned that 
`Focused Module: The ConfigResolver is dedicated to fetching and aggregating content. Content parsing is handled in the config-to-blueprint step, ensuring modularity.`
Since content parsing is not done in this module It doesn't know if the linked file/URL contains any additional links. If we consider this case, we need to parse the content and add the links to the `Vec<Link>`.

tusharmath replied:
`blueprint::from_config` is not async. So we can't perform any IOs in that step. Everything will need to be resolved before it reaches that step. We could parse the files and insert it in the config as data but not validate it. Parsing could be the responsibility of the Resolver module.

melsonic replied:
adding the unit test in the next commit :+1:

ologbonowiwi commented on src/config/resolver.rs:
```diff
@@ -0,0 +1,96 @@
+
+  async fn get_raw_content(link: &Link) -> anyhow::Result<Link> {
+    let mut res_link: Link = link.clone();
+    if let Ok(url) = Url::parse(&link.src) {
+      let resp = reqwest::get(url).await?;
+      if !resp.status().is_success() {
+        return Err(anyhow!("Read over URL failed with status code: {}", resp.status()));
+      }
+      res_link.content = Some(resp.text().await?);
+    } else {
+      let path = &link.src.trim_end_matches('/');
+      let mut f = File::open(path).await?;
+      let mut buffer: Vec<u8> = Vec::new();
+      f.read_to_end(&mut buffer).await?;
+      res_link.content = Some(String::from_utf8(buffer)?);
+    };
```
we could make the `if`/`else` block return the `content` then we override.

```suggestion
    let content = if let Ok(url) = Url::parse(&link.src) {
      let resp = reqwest::get(url).await?;
      if !resp.status().is_success() {
        return Err(anyhow!("Read over URL failed with status code: {}", resp.status()));
      }
      Some(resp.text().await?);
    } else {
      let path = &link.src.trim_end_matches('/');
      let mut f = File::open(path).await?;
      let mut buffer: Vec<u8> = Vec::new();
      f.read_to_end(&mut buffer).await?;
      Some(String::from_utf8(buffer)?);
    };
    
    // override and return
```

this would also avoid needing to create `res_link` as mutable to return it afterward.

ologbonowiwi commented on src/config/resolver.rs:
```diff
@@ -0,0 +1,96 @@
+        if !resp.status().is_success() {
+          return Err(anyhow!("Read over URL failed with status code: {}", resp.status()));
+        }
+        let source = if let Some(v) = resp.headers().get("content-type") {
+          if let Ok(s) = Source::detect_content_type(v.to_str()?) {
+            s
+          } else {
+            Source::detect(path.trim_end_matches('/'))?
+          }
+        } else {
+          Source::detect(path.trim_end_matches('/'))?
+        };
```
likely we can make the condition response an `Option` enum so we can make `let source = source.or(Source::detect(path.trim_end_matches('/'))?)` instead of returning `Source::detect(path.trim_end_matches('/'))?` on `else` blocks.

ologbonowiwi replied:
another option is to put this logic into a private function.

so you can make something like
```suggestion
        let source = get_source(resp).or(Source::detect(path.trim_end_matches('/'))?;
```

ologbonowiwi replied:
and the private function would be closer to this:
```
  fn get_source(resp: /* the type */) -> anyhow::Result<Source> {
    let content_type = resp.headers().get("content-type")?;

    Source::detect_content_type(content_type)
  }
```

note this is not a ready-to-go code, and it'll likely require a few tweaks

ologbonowiwi commented:
looks like there are some tests broken @melsonic. if you need a hand or want to collaborate on it, let me know, as #859 (which I'm assigned to) will rely on `@link` to load environment variables.

ologbonowiwi commented on src/config/resolver.rs:
```diff
@@ -0,0 +1,96 @@
+    let mut link_queue: VecDeque<Link> = VecDeque::new();
+    link_queue.push_back(self);
+    while let Some(mut curr) = link_queue.pop_front() {
+      let (txt, source) = if let Ok(url) = Url::parse(&curr.src) {
+        let resp = reqwest::get(url).await?;
+        let path = curr.src.clone();
+        if !resp.status().is_success() {
+          return Err(anyhow!("Read over URL failed with status code: {}", resp.status()));
+        }
+        let source = if let Some(v) = resp.headers().get("content-type") {
+          if let Ok(s) = Source::detect_content_type(v.to_str()?) {
+            s
+          } else {
+            Source::detect(path.trim_end_matches('/'))?
+          }
+        } else {
+          Source::detect(path.trim_end_matches('/'))?
+        };
+        (resp.text().await?, source)
```
this whole `if` condition block could be a private function. IMO, it'd make easier to read it.

ologbonowiwi commented on src/config/resolver.rs:
```diff
@@ -0,0 +1,96 @@
+        (String::from_utf8(buffer)?, Source::detect(path)?)
+      };
+
+      curr.content = Some(txt.clone());
+      let config = Config::from_source(source, &txt)?;
```
if you change the lines' order, you don't need to clone `txt`

```suggestion
      let config = Config::from_source(source, &txt)?;

      curr.content = Some(txt);
```

ologbonowiwi commented on src/config/resolver.rs:
```diff
@@ -0,0 +1,96 @@
+        };
+        (resp.text().await?, source)
+      } else {
+        let path = &curr.src.trim_end_matches('/');
+        let mut f = File::open(path).await?;
+        let mut buffer = Vec::new();
+        f.read_to_end(&mut buffer).await?;
+        (String::from_utf8(buffer)?, Source::detect(path)?)
```
these lines are repeated on `get_raw_content` else block. we could put it into a function to be re-used.

ologbonowiwi commented on src/config/resolver.rs:
```diff
@@ -0,0 +1,96 @@
+impl Link {
+  pub async fn resolve_recurse(self) -> anyhow::Result<Vec<Link>> {
+    let mut result: Vec<Link> = Vec::new();
+    if self.type_of == LinkType::Protobuf || self.type_of == LinkType::Data {
+      let link: Link = Self::get_raw_content(&self).await?;
+      result.push(link);
+      return Ok(result);
+    }
```
since all files can be both `http://` or `file://`, why we handle `Protobuf` and `Data` differently?

melsonic replied:
Since we don't need to detect the source type for these files. Because Config and GraphQL files could have recursive links, but these files don't. To, remove unnecessary processing I have created a different function for these two types

ologbonowiwi replied:
from my more preferred to least preferred, I'd say that we have a few alternatives:

1. load source every time and don't check for it at all, as it'll just not find a link to recursive resolve;
2. move this condition to lines 48 and 63, early returning in case it matches;
3. move the loading (of both file/network protocol) to specific functions that can be re-used in both cases;

in case we go with `2.`, feel free to use `get_raw_content` as much as we don't duplicate the code for loading stuff (so this method will need to rely on the new methods).

IMO: if the processing is not high costly, we can certainly load everything and early return in case is `Protobuf` or `Data`.

melsonic commented:
@ologbonowiwi  I've addressed the issues with the broken tests by making the `merge_right` function asynchronous, as the newly added `merge_links` function involves I/O operations. However, I am finding a problem with the unit test in the `resolver_module`. Unlike other attributes in the config, which are simple maps/structs, the links attribute stores a vector. It seems that we need to either modify the `from_directive` function or implement a workaround to handle this unique case.

melsonic commented:
> looks like there are some tests broken @melsonic. if you need a hand or want to collaborate on it, let me know, as #859 (which I'm assigned to) will rely on `@link` to load environment variables.

let's talk on discord :+1:

ologbonowiwi commented on src/config/config.rs:
```diff
@@ -210,6 +214,18 @@ fn merge_unions(
+  // let runtime = tokio::runtime::Runtime::new()?;
+  for link in other_links {
+    // let recurse_links = runtime.block_on(async { Link::resolve_recurse(link).await })?;
+    let recurse_links = Link::resolve_recurse(link).await?;
```
IMO, we should have one step to merge the configs and other step to actually load it.

here we should make only the sync operation of merging the data structure. the loading of it should be on other step.

tusharmath commented:
closing in favour of #911
